Fix build by xlc on AIX

ksromanov opened the pull request:
This fixes build by xlc compiler:

1. enables GCC-style `__attribute__((aligned(n)))` for `xlc` 
May be it worth enabling this branch by-default instead of reporting error, since it is supported by clang, icc, Sun cc and xlc.

2. disables #line directives generation by `xlc` in CPP mode (preprocessor). 
These line directives look like
`# line 8 "file.c"`
which is incompatible with Ocaml.
The only files, which are currently compiled by Ocaml after C-preprocessor are `domainstate.ml*`.  

The incompatibilities were introduced in PR #8713 and are both in 4.10 and trunk.

ksromanov commented on runtime/caml/misc.h:
```diff
@@ -96,7 +96,7 @@ CAMLdeprecated_typedef(addr, char *);
 /* Alignment */
-#ifdef __GNUC__
+#if defined(__GNUC__) || defined(__xlC__)
 #define CAMLalign(n) __attribute__((aligned(n)))
```
Actually this attribute is supported by almost any more-or-less recent UNIX C compiler (2010+).

dra27 replied:
It would be a good change - would you like to convert `configure.ac` to use [`AC_COMPILE_IFELSE`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/html_node/Running-the-Compiler.html#Running-the-Compiler). It would allow the error to be caught where it should be during `configure`

dra27 commented on configure.ac:
```diff
@@ -520,7 +520,8 @@ OCAML_CC_VENDOR
+    [outputobj='-o $(EMPTY)'; gcc_warnings="-qflag=i:i"; # all warnings enabled
+     CPP="$(CC) -E -qnoppline"], # suppress incompatible XLC line directives
   [msvc-*],
     [outputobj=-Fo; CPP="cl -nologo -EP"; gcc_warnings=""],
```
```suggestion
    [outputobj=-Fo; gcc_warnings=""],
```
(`CPP` is already for MSVC in 450-452)

ksromanov replied:
> It should be relatively simple for to get configure.ac to test whether __attribute((aligned... is available which would solve this permanently for exotic C compilers and also move the error detection out of the C headers and into configure where it belongs.

I'll try to do it.

dra27 commented on configure.ac:
```diff
@@ -520,7 +520,8 @@ OCAML_CC_VENDOR
   [xlc-*],
-    [outputobj='-o $(EMPTY)'; gcc_warnings="-qflag=i:i"], # all warnings enabled
+    [outputobj='-o $(EMPTY)'; gcc_warnings="-qflag=i:i"; # all warnings enabled
+     CPP="$(CC) -E -qnoppline"], # suppress incompatible XLC line directives
```
Could this instead be put in LL450-452 (where the MSVC-specific version of `CPP` is set as well?). The setting below is a mistake...

dra27 replied:
There's a large comment above that bit about why it's preferred to invoke the C pre-processor directly - is it possible to do that?

dra27 reviewed (commented):
`configure` needs to be regenerated. The code for setting `CPP` was accidentally in two places - please could you move it LL450 (and delete the clearly spurious double-setting of it for MSVC - CI will catch anything there).

It should be relatively simple for to get `configure.ac` to test whether `__attribute((aligned...` is available which would solve this permanently for exotic C compilers and also move the error detection out of the C headers and into `configure` where it belongs.

xavierleroy commented:
I'm not quite sure why this `CAMLalign` macro is so important.  It is used to 8-align the members of the `caml_domain_state` struct, but why should these members be 8-aligned even on a 32-bit platform?  Why not define `CAMLalign(n)` to nothing in the default case?  Maybe I missed a discussion.  

Independently, the C 2011 standard way to modify alignment is `_Alignas(8)`.  So,
```
#if __STDC_VERSION__ >= 201112L
#define CAMLalign(n) _Alignas(n)
#else  
```
might be a good idea.

Secondly, `__attribute__` is widely available but it not part of any C standard, which is why we can't use it blindly.

Finally, I'm skeptical that these tests are more easily done in configure.

ksromanov commented:
> I'm not quite sure why this `CAMLalign` macro is so important. It is used to 8-align the members of the `caml_domain_state` struct, but why should these members be 8-aligned even on a 32-bit platform?

I discussed this with former team member - it looks like this requires people from Multicore project, @kayceesrk .

> Independently, the C 2011 standard way to modify alignment is `_Alignas(8)`.
> Secondly, `__attribute__` is widely available but it not part of any C standard, which is why we can't use it blindly.

Thanks a lot!

kayceesrk commented:
There are a few doubles in the domain state. They consume 8-bytes even on a 32-bit platform. Easiest to align everything to 8-bytes rather than more macro magic. There will only ever be one domain state structure per domain. This was brought up in the original PR as well and here is the link to the discussion: https://github.com/ocaml/ocaml/pull/8713#discussion_r305885377.

kayceesrk commented:
I wasn't aware of `_Alignas(n)`. This should work for `CAMLalign(n)`:

```diff
diff --git a/runtime/caml/misc.h b/runtime/caml/misc.h
index 91457a303..9280b9817 100644
--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -96,7 +96,9 @@ CAMLdeprecated_typedef(addr, char *);
 #endif
 
 /* Alignment */
-#ifdef __GNUC__
+#if __STDC_VERSION__ >= 201112L
+#define CAMLalign(n) _Alignas(n)
+#elif defined(__GNUC__)
 #define CAMLalign(n) __attribute__((aligned(n)))
 #elif _MSC_VER >= 1500
 #define CAMLalign(n) __declspec(align(n))
```

ksromanov committed:
Add attribute(align()) detection and fix XLC build.

dra27 commented:
Given that there is still (just) the odd Unix C compiler which doesn't support C11, this could easily be combined with the proper probe for `__attribute__((aligned(n)))` and this is then permanently done.

If something is specific to GCC (or GCC+Clang), then it's reasonable to use `__GNUC__` but if it's also in other compilers, then better to do the detection properly in `configure`, rather than add a potentially ever-growing list of macros to the `#if ...`. We already do quite a few of those, this one's also straightforward:

```diff
diff --git a/aclocal.m4 b/aclocal.m4
index ff12869a29..5ac1b7293e 100644
--- a/aclocal.m4
+++ b/aclocal.m4
@@ -86,6 +86,14 @@ AC_DEFUN([OCAML_CC_HAS_FNO_TREE_VRP], [
   CFLAGS="$saved_CFLAGS"
 ])
 
+AC_DEFUN([OCAML_CC_SUPPORTS_ALIGNED], [
+  AC_MSG_CHECKING([whether the C compiler supports __attribute__((aligned(n)))])
+  AC_COMPILE_IFELSE(
+    [AC_LANG_SOURCE([typedef struct {__attribute__((aligned(8))) int t;} t;])],
+    [AC_DEFINE([SUPPORTS_ALIGNED_ATTRIBUTE])
+    AC_MSG_RESULT([yes])],
+    [AC_MSG_RESULT([no])])])
+
 AC_DEFUN([OCAML_CC_HAS_DEBUG_PREFIX_MAP], [
   AC_MSG_CHECKING([whether the C compiler supports -fdebug-prefix-map])
   saved_CFLAGS="$CFLAGS"
diff --git a/configure b/configure
index fa4787983d..0b30e9b139 100755
Binary files a/configure and b/configure differ
diff --git a/configure.ac b/configure.ac
index 63e2354ee3..da431e33ad 100644
--- a/configure.ac
+++ b/configure.ac
@@ -843,6 +843,8 @@ AS_CASE(["$CC,$host"],
     AS_IF([$cc_has_fno_tree_vrp],
       [internal_cflags="$internal_cflags -fno-tree-vrp"])])
 
+OCAML_CC_SUPPORTS_ALIGNED
+
 # Configure the native-code compiler
 
 arch=none
diff --git a/runtime/caml/m.h.in b/runtime/caml/m.h.in
index 1a1575db9b..4299643aa0 100644
--- a/runtime/caml/m.h.in
+++ b/runtime/caml/m.h.in
@@ -104,3 +104,5 @@
 #undef FLAT_FLOAT_ARRAY
 
 #undef FUNCTION_SECTIONS
+
+#undef SUPPORTS_ALIGNED_ATTRIBUTE
diff --git a/runtime/caml/misc.h b/runtime/caml/misc.h
index 91457a303c..3ec0d6a72f 100644
--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -96,7 +96,9 @@ CAMLdeprecated_typedef(addr, char *);
 #endif
 
 /* Alignment */
-#ifdef __GNUC__
+#if __STDC_VERSION__ >= 201112L
+#define CAMLalign(n) _Alignas(n)
+#elif defined(SUPPORTS_ALIGNED_ATTRIBUTE)
 #define CAMLalign(n) __attribute__((aligned(n)))
 #elif _MSC_VER >= 1500
 #define CAMLalign(n) __declspec(align(n))
```

stedolan commented:
Some more context on the alignment stuff: The code generated by `ocamlopt` makes direct references into the domain state structure, which is stored in a register on many platforms (e.g. `%r14` on amd64). For instance, the allocation hotpath reads the `young_limit` field from this structure. For this to work, we need to be able to compute the exact offset of each member.

C structure layout rules can be tricky, so the simplest way to accomplish this is to force C to lay out the structure by giving each member exactly 8 bytes, so that the offsets [can be computed easily](https://github.com/ocaml/ocaml/blob/trunk/asmcomp/amd64/emit.mlp#L162-L163). (There's [some magic](https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/domain_state.h#L37-L49) in `domain_state.h` to assert at compile time that the C compiler did as we asked).

(The padding this introduces wastes a few dozen bytes on 32-bit platforms, but since there's only one of the structures in the whole VM it doesn't matter)

ksromanov commented:
Thanks a lot for the comments!

Updates:

1. Added @dra27 patch to configure.ac . Since I don't have MSVC, I couldn't test on compiler which does not support `__attribute__((aligned(n)))`. `Xlc`, `Sun C` do support this attribute.

2. Fixed code setting `CPP` - it is now set around LL450. Since `CPP` depends on the compiler, but not on the `host`, I had to move `OCAML_CC_VENDOR` macro above LL450.

3. Updated `CAMLalign` macro.

4. Tested on 4 compilers - `xlc` 12.1, `Sun C` 5.12, `gcc` 7.3.1, `clang` 9.0.0:
`xlc` and `Sun C` use `__attribute__((aligned(n)))`, while `gcc` and `clang` use `C 2011` standard.

`Sun C` currently fails compilation phase by different reasons.

_Question:_ do we need fallback `#define CAMLalign(n)` for bytecode version? As far as I understood:

1. Almost all available compilers either support `__attribute__((aligned(n)))` or `C 2011` standard, or it is MSVC.

2. Bytecode version actually does not need `CAMLalign(n)`, so for better compatibility we can add 
```C
#elif !defined(NATIVE_CODE)
#define CAMLalign(n)
```

kayceesrk commented:
Along with the proposed macro definition for bytecode-only version, you will have to suitably wrap https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/domain_state.h#L37-L49 with a conditional check.

ksromanov commented:
Thanks, @kayceesrk !

So, since all the compilers I've ever used, are now covered by either C 2011, `__declspec(align(n))` or `__attribute__((aligned(n)))`; and support for empty `CAMLalign` needs changes in https://github.com/ocaml/ocaml/blob/trunk/runtime/caml/domain_state.h#L37-L49 
may be we should stop at the current state of this PR?

Or it still worth adding empty `CAMLalign` for bytecode-only targets? (we can not reliably test it)

shindere commented:
To me the PR looks good in its current state.

Could you just squash the commits, please?

I don't think we should go further here.

ksromanov commented:
Rebased + squashed.

Does this PR need CHANGES entry, since David (@dra27) added `attribute(aligned())` detection to `configure.ac`? The rest is just a bug-fix update.

shindere commented:
Thanks a lot, @ksromanov. I went ahead and merged although CI was not
green, I suspect it's because the PR ahs neither a Changes entry nor a
`no-change-entry-needed` label but I think this is okay.

Am I correct that this should be cherry-picked to the4.10 branch?

ksromanov commented:
Thanks a lot, Sébastien!

It would be desirable to cherry-pick these changes into 4.10, but this branch will have merge conflicts 4.10 branch in configure. I have two branches in my fork `xlc_update_4.10_unsquashed` (rebased on 4.10 before squash + replaced configure regeneration commit) and `xlc_update_4.10` (squashed, rebased on 4.10). I hope these branches will help.

shindere commented:
Konstantin Romanov (2019/11/25 12:55 -0800):
> Thanks a lot, Sébastien!

My pleasure :)

> It would be desirable to cherry-pick these changes into 4.10,

Done in 0345fda28d876822b179edd72d460298e3b20c7b

> but this branch will have merge conflicts 4.10 branch in configure.

These conflicts are actually easily solved by re-generating configure
from the configure.ac.

ksromanov commented:
Thanks a lot!
